### PR TITLE
4/4 Improve Simulation class flexibility

### PR DESCRIPTION
**Pull request by arbennett**
This follows #58 which should be merged first.

This adds several features to the Simulation class.  First, it differentiates between running and executing, which means that processing can be done without halting a workflow (making SUMMA run in the background).  Second it stores standard and error outputs in properties, making it easier to debug SUMMA runs. It also has a status property for quickly checking the state of the object.  Finally, this provides cleaner interoperability with the configuration refactor, making it easier to get output files.

### CODE REVIEW

File: pysumma/simulation.py [hunk 1]
```diff
@@ -39,161 +39,165 @@ def __init__(self, executable, filemanager, run_suffix='_pysumma_run'):
...
-                    subprocess.run(merge_netcdf, shell=True)
-                    myList = ' '.join(output_list[:])
-                    delete_netcdf = 'rm -rf ' + myList
-                    subprocess.run(delete_netcdf, shell=True)
+    def exec_hydroshare(self, run_suffix, specworker_img=None):
+        #TODO: This needs to be updated
+        raise NotImplementedError('This needs to be updated')
+        self.run_suffix = ""
+        #TODO: This should be up top
+        from specworker import jobs
```
Comment (DavidChoi76) [inline pysumma/simulation.py]: This command works only in hydroshare execution. If I move this command at the top, the error occurs.

File: pysumma/simulation.py [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+            if o.period == 1:
+                output_files.add(base_file.format('timestep'))
+            elif o.period > 1:
+                output_files.add(base_file.format(o.statistic))
+        return list(output_files)
+
+    def execute(self, run_option, run_suffix=None,
+                preprocess_cmds=[], monitor=False):
+        """Run a SUMMA simulation"""
+        self.start(run_option, run_suffix, preprocess_cmds)
```
Comment (DavidChoi76) [inline pysumma/simulation.py:208]: When I ran SUMMA results_simpleResistance = S.execute(run_option = 'docker_sopron_2018'), there is an error like AttributeError: 'OutputControlOption' object has no attribute 'value'. When I open Model_Output.txt, there was no variable. Do I have to define variable? What is the problem?

```---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-8-13791bfb579d> in <module>()
----> 1 results_simpleResistance = S.execute(run_option = 'docker_sopron_2018')

~/miniconda3/lib/python3.6/site-packages/pysumma/simulation.py in execute(self, run_option, run_suffix, preprocess_cmds, monitor)
    176                 preprocess_cmds=[], monitor=False):
    177         """Run a SUMMA simulation"""
--> 178         self.start(run_option, run_suffix, preprocess_cmds)
    179         if monitor:
    180             result = self.monitor()

~/miniconda3/lib/python3.6/site-packages/pysumma/simulation.py in start(self, run_option, run_suffix, preprocess_cmds)
    107         errstring = ('No executable defined. Set as "executable" attribute'
    108                      ' of Simulation or check run_option ')
--> 109         self._write_configuration()
    110         if run_suffix:
    111             self.run_suffix = run_suffix

~/miniconda3/lib/python3.6/site-packages/pysumma/simulation.py in _write_configuration(self)
    159         self.local_param_info.write()
    160         self.basin_param_info.write()
--> 161         self.output_control.write()
    162 
    163     def _get_output(self):

~/miniconda3/lib/python3.6/site-packages/pysumma/option.py in write(self, path)
     73         with open(path, 'w') as f:
     74             f.writelines(self.header)
---> 75             f.writelines((repr(o) + '\n' for o in self.options))
     76 
     77     def get_option(self, name, strict=False):

~/miniconda3/lib/python3.6/site-packages/pysumma/option.py in <genexpr>(.0)
     73         with open(path, 'w') as f:
     74             f.writelines(self.header)
---> 75             f.writelines((repr(o) + '\n' for o in self.options))
     76 
     77     def get_option(self, name, strict=False):

~/miniconda3/lib/python3.6/site-packages/pysumma/option.py in __repr__(self)
     14 
     15     def __repr__(self):
---> 16         return "{} : {}".format(self.name, self.value)
     17 
     18 

AttributeError: 'OutputControlOption' object has no attribute 'value'```

File: pysumma/simulation.py [hunk 3]
```diff
@@ -39,161 +39,165 @@ def __init__(self, executable, filemanager, run_suffix='_pysumma_run'):
...
-                        + self.run_suffix + '1.nc')
-                    merge_netcdf = ('ncrcat ' + myoutput_list
-                                    + ' -O ' + output_name)
-                    subprocess.run(merge_netcdf, shell=True)
-                    myList = ' '.join(output_list[:])
-                    delete_netcdf = 'rm -rf ' + myList
-                    subprocess.run(delete_netcdf, shell=True)
+    def exec_hydroshare(self, run_suffix, specworker_img=None):
+        #TODO: This needs to be updated
+        raise NotImplementedError('This needs to be updated')
```
Comment (DavidChoi76) [inline pysumma/simulation.py]: I don't know what this command mean? Do we need raise NotImplementedError here?

File: pysumma/simulation.py [hunk 4]
```diff
@@ -39,161 +39,165 @@ def __init__(self, executable, filemanager, run_suffix='_pysumma_run'):
...
+            myoutput_list = ' '.join(output_list[1:])
+            if myoutput_list.count(self.output_prefix.value) > 1:
+                output_name = (
+                    self.base_dir + self.manager.output_path.value.split('>')[1]
                     + self.output_prefix.value + '_'
                     + self.decision_obj.simulStart.value[0:4] + '-'
-                    + self.decision_obj.simulFinsh.value[0:4] + '_' + '1.nc')
+                    + self.decision_obj.simulFinsh.value[0:4] + '_'
+                    + self.run_suffix + '1.nc')
+                #TODO: This could be done via xarray - also eneralized
```
Comment (DavidChoi76) [inline pysumma/simulation.py]: I will try to change later. But now we don't use the master version of summa anymore. So this is not a big deal.

File: pysumma/simulation.py [hunk 5]
```diff
@@ -39,161 +39,165 @@ def __init__(self, executable, filemanager, run_suffix='_pysumma_run'):
...
+                merge_netcdf = ('ncrcat ' + myoutput_list
+                                + ' -O ' + output_name)
+                subprocess.run(merge_netcdf, shell=True)
+                myList = ' '.join(output_list[:])
+                #TODO: This is a bad idea
```
Comment (DavidChoi76) [inline pysumma/simulation.py]: I will consider this later when I apply xarray.

File: pysumma/simulation.py [hunk 6]
```diff
@@ -39,161 +39,165 @@ def __init__(self, executable, filemanager, run_suffix='_pysumma_run'):
...
-                out_file_path = (
-                    self.base_dir + '/'
-                    + self.output_path.filepath.split('/')[1] + '/'
-                    + self.output_prefix.value + '_output_' + 'timestep.nc')
+        elif specworker_img == 'cuahsi/summa:sopron':
+            out_file_path = (
+                self.base_dir + '/'
+                + self.output_path.filepath.split('/')[1] + '/'
+                + self.output_prefix.value + '_output_' + 'timestep.nc')
+            #TODO: WHy doesn't this run anything?
```
Comment (DavidChoi76) [inline pysumma/simulation.py]: running command is 91 line.  This process is for assigning output_path.

File: pysumma/simulation.py [hunk 7]
```diff
@@ -39,161 +39,165 @@ def __init__(self, executable, filemanager, run_suffix='_pysumma_run'):
...
+        else:
+            #TODO: This error message isn't specific enough
+            raise ValueError('You need to deinfe the '
+                             'exact SUMMA_image_name')
```
Comment (DavidChoi76) [inline pysumma/simulation.py:133]: I want to change like this.
`You need to check specworker_img name. Your image name must be cuahsi/summa:sopron or cuahsi/summa:master.`

File: pysumma/simulation.py [hunk 8]
```diff
@@ -39,161 +39,165 @@ def __init__(self, executable, filemanager, run_suffix='_pysumma_run'):
...
+        return out_file_path
 
-            else:
-                raise ValueError('You need to deinfe the '
-                                 'exact SUMMA_image_name')
+    def exec_docker(self, run_suffix, docker_img):
+
+        self.executable = docker_img
```
Comment (DavidChoi76) [inline pysumma/simulation.py]: this needs to be change: self.executable = 'uwhydro/summa:sopron_2018'

File: pysumma/simulation.py [hunk 9]
```diff
@@ -39,161 +39,165 @@ def __init__(self, executable, filemanager, run_suffix='_pysumma_run'):
...
+    def exec_hydroshare(self, run_suffix, specworker_img=None):
+        #TODO: This needs to be updated
+        raise NotImplementedError('This needs to be updated')
+        self.run_suffix = ""
+        #TODO: This should be up top
+        from specworker import jobs
+        # define the image that we want to execute
+        # save these paths in the env_vars dictionary
+        # which will be passed to the model
+        env_vars = {'LOCALBASEDIR': self.base_dir,
```
Comment (DavidChoi76) [inline pysumma/simulation.py]: From here to end, I need to change self.base_dir, self.output_path ... etc considering this pull request.
For example, self.output_path.value = self.manager.output_path.value

File: pysumma/simulation.py [hunk 10]
```diff
@@ -39,161 +39,165 @@ def __init__(self, executable, filemanager, run_suffix='_pysumma_run'):
...
+                + run_suffix + '_timestep.nc')
+        return out_file_path
 
+    def execute(self, run_option, run_suffix, specworker_img=None):
+        # set run_suffix to distinguish the output name of summa
+        self.run_suffix = run_suffix
+        # 'local' run_option runs summa with summa execution
+        # file where is in a local computer.
+        if run_option == 'local':
+            out_data = self.execute_local(run_suffix)
```
Comment (DavidChoi76) [inline pysumma/simulation.py]: need to change self.execute_local(run_suffix) to self.exec_local(run_suffix)

File: pysumma/simulation.py [hunk 11]
```diff
@@ -39,161 +39,165 @@ def __init__(self, executable, filemanager, run_suffix='_pysumma_run'):
...
+        # 'docker_sopron_2018' run_option runs summa with docker hub online,
+        # and the version name is "'uwhydro/summa:sopron_2018'.
+        elif run_option == "docker_sopron_2018":
+            out_data = self.execute_docker(run_suffix, 'docker_sopron_2018')
```
Comment (DavidChoi76) [inline pysumma/simulation.py]: need to change self.execute_docker(run_suffix, 'docker_sopron_2018') to self.exec_docker(run_suffix)

File: pysumma/simulation.py [hunk 12]
```diff
@@ -39,161 +39,165 @@ def __init__(self, executable, filemanager, run_suffix='_pysumma_run'):
...
+        # "specworker" run_option run summa with summa image
+        # in docker of HydroShare Jupyter Hub
+        elif run_option == "specworker":
+            out_data = self.execute_local(run_suffix,
```
Comment (DavidChoi76) [inline pysumma/simulation.py]: need to change self.execute_local(run_suffix, to self.exec_local(run_suffix,

File: pysumma/simulation.py [hunk 13]
```diff
@@ -39,161 +39,165 @@ def __init__(self, executable, filemanager, run_suffix='_pysumma_run'):
...
+        else:
+            #TODO: This error message isn't specific enough
+            raise ValueError('You need to deinfe the '
+                             'exact SUMMA_image_name')
+        return out_file_path
 
-            else:
-                raise ValueError('You need to deinfe the '
-                                 'exact SUMMA_image_name')
+    def exec_docker(self, run_suffix, docker_img):
```
Comment (DavidChoi76) [inline pysumma/simulation.py]: need to change exec_docker(self, run_suffix, docker_img) to exec_docker(self, run_suffix):

File: pysumma/simulation.py (hunk 2 above)
Comment (arbennett) [reply pysumma/simulation.py:208]: This should now be updated and fixed.

Comment (DavidChoi76) [reply pysumma/simulation.py:208]: I think there is no problem in this part. The actual error was the self.executable = docker_img at exec_docker method. So we need to change to self.executable = 'uwhydro/summa:sopron_2018'

Comment (arbennett) [reply pysumma/simulation.py:208]: The actual error was `AttributeError: 'OutputControlOption' object has no attribute 'value'` - which was caused by my accidental misnaming of the `__repr__` function in the `OutputControlOption` class. 

What you've said may be an additional error though.

Comment (DavidChoi76) [reply pysumma/simulation.py:208]: Ok. But when I ran `S.execute(run_option = 'docker_sopron_2018')`, there were some problems in my pysumma code. I guess I used pysumma code that applied until pull request 3. So I updated pysumma code until pull request 4 and fixed self.executable. Now I don't have any error here when I ran `S.execute(run_option = 'docker_sopron_2018', run_suffix="simpleResistance").

**Pull request merged**